### ISSUE TITLE
Azure Bot gets on turn error after 64 turn / step  : The reader's Max Depth of 64 has been exceeded

### ISSUE THREAD

**Issue by AsmaaHamedEwes** [microsoft/BotBuilder-Samples#3328]
[transcripts.zip](https://github.com/microsoft/BotBuilder-Samples/files/6688493/transcripts.zip)
```
{
  "type": "trace",
  "timestamp": "2021-06-21T17:34:16.738Z",
  "serviceUrl": "http://localhost:29562",
  "channelId": "emulator",
  "from": {
    "id": "bab77c40-d2b6-11eb-b628-ef359bd77808",
    "name": "Bot",
    "role": "bot"
  },
  "conversation": {
    "id": "bac31500-d2b6-11eb-bdc5-95488adbb945|livechat"
  },
  "recipient": {
    "id": "32c1f11b-649e-471c-91bd-87ce18d2b689",
    "role": "user"
  },
  "locale": "en-US",
  "replyToId": "e3f1d650-d2b6-11eb-bdc5-95488adbb945",
  "label": "TurnError",
  "valueType": "https://www.botframework.com/schemas/error",
  "value": ". Path 'DialogState.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.options.Prompt.attachments.$values'.",
  "name": "OnTurnError Trace",
  "id": "e70fb820-d2b6-11eb-bdc5-95488adbb945",
  "localTimestamp": "2021-06-21T19:34:16+02:00"
}
```

even if after each dialog I call      await stepContext.EndDialogAsync();

**Comment by stevkan** [comment]
@AsmaaHamedEwes, can you provide the code and explain what is happening prior to this trace activity? The value in the `value` property is appearing to be the problem or, at least, part of the problem. I'm curious to know how that is being generated.

Also, the 'transcript.zip' file you included is either empty or corrupted. After downloading, it produces an error when trying to open.

**Comment by AsmaaHamedEwes** [comment]
I created a customized core bot that has a set of Journeys starting with welcome Journey, then display user menu with a set of options each option triggers a separate dialog flow when any user input comes it  is being classified and detected by LUIS Dispatch then decide to continue this journey or move to chitchat journey . Keep in mind,  I am using bot builder SDK version 4.11.1 for backend development , Azure Bot APP service for hosting my bot and managing errors via app insights logging. 
I will attach the transcript file again  
[transcripts.zip](https://github.com/microsoft/BotBuilder-Samples/files/6696228/transcripts.zip)
you need to have Bot Framework Emulator installed to open transcript file

**Comment by AsmaaHamedEwes** [comment]
any advice @mrivera-ms

**Comment by hcyang** [comment]
Hi @AsmaaHamedEwes, I tried to download and open the attached transcripts.zip file, but somehow it is not a valid ZIP file and only 9KB in length. Please verify if you had uploaded the correct file and maybe re-upload it again. Thanks.

**Comment by AsmaaHamedEwes** [comment]
Hi , @hcyang , I will try to Upload it again. but let me rephrase the Issue again as I got some notes while I was debugging . 
first while investigating I found that working turn context has 9 items in  turn state elements , error  turn  context has 4 turn state items only exactly fails at saving the conversation conversation state. bot builder uses the cached bot state to save to it the conversation state  in each turn bot builder add the turn state to bot conversation state as following :
 Conversation state => state => [0] => Dialog state => Dialog stack => Main Dialog => state => [0] => dialogs => Dialog stack => main menu dialog => state => dialogs  => Dialog stack => ….. and so on for all dialogs with in conversation.      
 at some point while try to save state with depth exceeded 64 limit it the mentioned exception is being thrown. 
I will attach screen shoot for my mentioned investigations .
![savingStateSnap](https://user-images.githubusercontent.com/50184240/125184687-273be680-e220-11eb-8543-1790f03b53f2.png)
![workingTurn](https://user-images.githubusercontent.com/50184240/125184689-299e4080-e220-11eb-8967-5409dbd86148.png)
![errorTurn](https://user-images.githubusercontent.com/50184240/125184769-c9f46500-e220-11eb-9c13-fbeb8728d026.png)
[transcripts.zip](https://github.com/microsoft/BotBuilder-Samples/files/6796436/transcripts.zip)
![Size of transcripts](https://user-images.githubusercontent.com/50184240/125184818-26578480-e221-11eb-8745-448b068071d0.png)

**Comment by tonyanziano** [comment]
@hcyang do you have an update on this issue?

**Comment by benbrown** [comment]
Have we determined whether or not this is a problem inherent in the SDK, or a behavior caused by this specific implementation?

**Comment by davidefu** [comment]
Hi,
I've the same issue and I've resolved downgrading to Newtonsoft json 12.0.3
I thing that due to https://github.com/JamesNK/Newtonsoft.Json/issues/2501 and the fix https://github.com/JamesNK/Newtonsoft.Json/pull/2504 introduced the error

the fix has been done here https://github.com/JamesNK/Newtonsoft.Json/pull/2505 we have to wait for a new version

**Comment by skommireddi** [comment]
> Hi , @hcyang , I will try to Upload it again. but let me rephrase the Issue again as I got some notes while I was debugging .
> first while investigating I found that working turn context has 9 items in turn state elements , error turn context has 4 turn state items only exactly fails at saving the conversation conversation state. bot builder uses the cached bot state to save to it the conversation state in each turn bot builder add the turn state to bot conversation state as following :
> Conversation state => state => [0] => Dialog state => Dialog stack => Main Dialog => state => [0] => dialogs => Dialog stack => main menu dialog => state => dialogs => Dialog stack => ….. and so on for all dialogs with in conversation.
> at some point while try to save state with depth exceeded 64 limit it the mentioned exception is being thrown.
> I will attach screen shoot for my mentioned investigations .
> ![savingStateSnap](https://user-images.githubusercontent.com/50184240/125184687-273be680-e220-11eb-8543-1790f03b53f2.png)
> ![workingTurn](https://user-images.githubusercontent.com/50184240/125184689-299e4080-e220-11eb-8967-5409dbd86148.png)
> ![errorTurn](https://user-images.githubusercontent.com/50184240/125184769-c9f46500-e220-11eb-9c13-fbeb8728d026.png)
> [transcripts.zip](https://github.com/microsoft/BotBuilder-Samples/files/6796436/transcripts.zip)
> ![Size of transcripts](https://user-images.githubusercontent.com/50184240/125184818-26578480-e221-11eb-8745-448b068071d0.png)

Hi @AsmaaHamedEwes I tried to open the latest transcript file you attached and it still says it is invalid. Can you try to send me a valid transcript file. Thanks

**Comment by skommireddi** [comment]
I was able to reproduce the issue using one of the existing samples "ComplexDialogBot"and modifying code to create multi-turn conversation. The MaxDepth error occurs only when targeting Newtonsoft.Json version 13.0.1. The same issue does not occur with 12.0.3 version of Newtonsoft.Json

![Json Max Depth error](https://user-images.githubusercontent.com/30093161/130529836-24c37984-c5a0-442c-9f5d-57d17f77cc4c.png)

Hence we suggest to downgrade to 12.0.3 version of Newtonsoft.Json until the latest version of Newtonsoft.Json is released with the fix for this error. JamesNK/Newtonsoft.Json#2505 fixes the error.

**Comment by hypertrends** [comment]
Hello, while this issue has been resolved by downgrading to Newtonsoft 12.0.3 as recommended, I happen to think that this might be a bigger issue down the line. It feels like the state stack isn't getting replaced/swapped out and each iteration subsequently adds more and more values to the stack - these are just repeated dialogs that are being replaced using call below. 

`stepContext.ReplaceDialogAsync`

Ideally, this should probably clear-off the stack and start from the top again, but it looks like it just keeps on chaining.

To put this in perspective, here's the actual serialization of state:

`Newtonsoft.Json.JsonReaderException: The reader's MaxDepth of 64 has been exceeded. Path 'DialogState.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.dialogs.dialogStack.$values[0].state.options.Prompt.attachments.$values[0].content.actions[0].data'.`

**Comment by crismogram** [comment]
Is this possibly related to https://github.com/microsoft/BotFramework-Emulator/issues/2400

How can I downgrade to Newtonsoft 12.0.3, I don't think I can do it on Package Manager.
